### DISCUSSION TITLE
Record response every time an image appears in one trial

### DISCUSSION THREAD

**Discussion by avashipman** [jspsych/jsPsych#945] [Q&A] [answered]
Hi,

I have an experiment that has on the top of the screen, a survey text and on the bottom of the screen, an automatic interval of images (which Vijay kindly helped me out on! 🤩) that has a target image in the interval which requires a keyboard response. So the user is typing a response while also reacting to the target image in the interval. 

I tried isolating the responses so that only the target image gets a correct_response ('8') while the other images get jsPsych.ALL_KEYS (since the person is typing) but '8' is only recorded once even though the target image appears multiple times in the interval:
https://github.com/avashipman/Survey-Distractor/blob/b1aa8363909aabc3ce13abe07413394002ae06d8/Alternate.html#L64-L77

I also made the plugin read that the correct response would be the trial.choices so that it wouldn't just read the first key pressed as the response:
https://github.com/avashipman/Survey-Distractor/blob/b1aa8363909aabc3ce13abe07413394002ae06d8/jspsych-survey-distractor.js#L307-L311

What would be ideal is if the data would show that every time the target image emerged, the choices key was pressed. However, I am not sure how to go about this. Also, I tried making the textbox input event restrict the use of numbers but nothing I've tried worked. Here is what I tried so far:    
 ```
//No numbers in textbox function (NOT WORKING: says cannot read property 'replace' of undefined) 
    function lettersOnly(input) {
      var regex = /[0-9]/g;
      input.value = input.value.replace(regex, "");
  } 
```
I would be so thankful for any advice on how to restrict the input of the textbox! 

Thank you for your time and I would appreciate any help you can give!!

**Comment by vijaymarupudi** [comment, marked as answer]
```
response.key == trial.choices
```

In this snipping, you are comparing a string to an Array, which will not provide valid output. Maybe you want `response.key == trial.choices[0]`?

lettersOnly is along the right path! However, I would not set events in the html like you did. Add a unique id to the input element, use `document.querySelector` to acquire a reference, and then use `addEventListener('event-name', <callback>)` to attach to the input.

This might help! https://javascript.info/introduction-browser-events#addeventlistener

**Reply by avashipman** [reply]
Thank you for the advice! 

I think the problem with the response is that the computer thinks that the interval of images is actually one class so that it doesn't receive other responses. When I press '8' the whole odd-ball task is considered responded instead of just that particular image being responded to. I tried changing the id so that each image would have a unique id, thus requiring a response of its own, but it doesn't work. How can I give a unique id to every time an image appears or assign the responded class to each appearing image ([L305-311](https://github.com/avashipman/Survey-Distractor/blob/612f2c8ed02cf5b0fccb765496c48b63bac9d4d9/jspsych-survey-distractor.js#L305-L311))? Should I put the array of images in the plugin itself?

Also, the input event still didn't work because it cannot read property 'replace' of undefined. But I think that the addEventListener is executing the function correctly, it's just the function that isn't working. Why would input.value be undefined?

**Reply by vijaymarupudi** [reply]
I would post updated code with `addEventListener` to troubleshoot further, I do not know how you've integrated it with your plugin. It might be that input is something else (I would use console.log to verify).

Just to clarify, you will display an array of images, and when the participant types a number, the current image could change, and the key used to respond should be stored, correct? Conceptually, this would be accomplished in the after_response callback, by pushing to a trial-global array called responses instead of assigning the response variable like I see it currently done.

**Reply by avashipman** [reply]
Here is the updated code with the addEventListener:
`  var handleClick = function (input) {
      var regex = /[0-9]/g;
      input.value = input.value.replace(regex, "");
    }; 
    document.querySelector("#survey-input")           
    addEventListener('input', handleClick)
`
I also tried:
`    function lettersOnly(input) {
      var regex = /[0-9]/g;
      input.value = input.value.replace(regex, "");
  }
    document.querySelector("#survey-input")
    addEventListener('input', lettersOnly(this))`

For the display of images, they are automatically shown in an interval, like a slideshow, and every time the target image appears from the interval I want the user to press "8". The target image appears several times while the trial is happening, so I want the response to be recorded every time the target image appears. Does that help to clarify?

**Reply by vijaymarupudi** [reply]
One thing that's important in programming are the types of the data you're dealing this, and really focusing on them as a beginner is what I typically recommend people when they are learning programming. Most bugs are often because the types are not the types one expects!

```javascript
document.querySelector("#survey-input")
addEventListener('input', lettersOnly(this))
```


In this case, you are calling a global function called `addEventListener`, which does not exist. You are missing a dot between the two lines.

Additionally, I presume the `this` came from copy-pasting code from elsewhere. `this` is a concept you do not need most of the time. This is probably carryover from converting your HTML string concatentation to pure javascript. [Element.addEventListener()] takes a function as the second argument. However, you are providing it `undefined`. You are calling the lettersOnly function at that line of code, and using the result for `addEventListener`. `lettersOnly` does not return anything, so this code is buggy.

What you probably wanted to do was pass in the lettersOnly function, like this:

```javascript
document.querySelector("#survey-input").addEventListener('input', lettersOnly)
```

Now in `lettersOnly`, the argument you are receiving is not the input, but an event object that contains metadata on the input ([more info here]). The `target` property of the object is the element that raised the event, which is the input. You'll need something like this.

```javascript
    function lettersOnly(event) {
      var regex = /[0-9]/g;
      var input = event.target;
      input.value = input.value.replace(regex, "");
    }
```

[more info here]: https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/input_event

[Element.addEventListener()]: https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener

Let's get this resolved first, before tackling the image interval problem. Could you outline what should happen if the participant does not respond with a number in the interval?

**Reply by avashipman** [reply]
Thank you for the explanation! When I tried the code you suggested I got Uncaught TypeError: Cannot read property 'addEventListener' of null. I made sure the ids were the same and that the placement of the code wasn't an issue, but I still got this message.
Do you think it might be because of the way the input tag is arranged ([171-182](https://github.com/avashipman/Survey-Distractor/blob/f741530170340c61f7cb378efcdcf87ff5a111e8/jspsych-survey-distractor.js#L171-L182))? I tried changing this as well, but maybe I am not doing it right.

For the interval of images, if the participant doesn't respond, the image still changes. So the interval keeps going like a carousel regardless of the participant's response.

**Reply by vijaymarupudi** [reply]
Right, that's because you're trying to query for an element that isn't on the page yet. Move the code down to after it has been assigned to the innerHTML of an element on the page.

In that case, what if a participant responds? Does the image switch immediately or does it wait for the interval?

**Reply by avashipman** [reply]
Ok, I was trying to place it after the htmlSurvey but the innerHTML makes sense! However, now the message reads: Cannot read property 'replace' of undefined at HTMLInputElement.lettersOnly.
It must not understand input.value because the error only happens when I press numbers, so when the function is actually being called.

When there is a response, the image waits for the interval.

**Reply by vijaymarupudi** [reply]
You didn't seem to implement the changes I suggested for lettersOnly?

**Reply by avashipman** [reply]
I am so sorry I forgot to include: var input = event.target;
I thought I included everything, but it works now!

**Reply by vijaymarupudi** [reply]
So a few notes on your image changing `setInterval` function, which will include some edits. It will run indefinitely, even after the trial is over. We will need to tell it to stop when the trial is over. We need to capture the interval id it returns and use it later.

```javascript
const intervalId = setInterval(function() {
  // ...
}, 1000)


// later

clearInterval(intervalId)
jsPsych.finishTrial()
```

Given that note, what would you like to happen when participants type numbers twice or thrice before the image changes? Do you plan on providing feedback to the participants?

Additionally, could you outline what data you'd like collected? Providing a pseudo JSON structure can help, something like this?


```javascript
[
  { "image": "a.jpg", "response": 8, rt: 3434 },
  // ...
]
```

Just a note, pushing when you reply helps me keep track of the changes you make!

**Reply by avashipman** [reply]
Sorry for not updating my changes when I reply. I will make sure to do that from now on. Thank you for letting me know! 

I hope that only the number '8' will be recorded when the target image appears and if the participant presses other numbers then it wouldn't be recorded and the images will still move on regardless of response. If the participant presses '8' for other images besides the target image, it doesn't have to be recorded. I just want to make sure that the participant is being distracted by responding correctly to the target image. The trial only ends when the participant clicks the submit button from the survey, so the interval of images won't determine the end of the trial. I am planning to not give any feedback to the participants.

I am thinking that the data could look like this:
```
[
  { "responses: "{\"Q0\":\"text that user types\"}",
  {"time_elapsed": 1234},
  {"image": "target.png", "response": 8, "correct_response": true}
  {"image": "target.png", "response": 8, "correct_response": true}  //if target image appears again in trial
  {"image": "target.png", "response": null, "correct_response": false}  //if participant doesn't press 8
  // ...
]

```
Something like this so that I can see that every time the target image appeared, the participant responded.
It can look like this as well:

```
[
  { "responses: "{\"Q0\":\"text that user types\"}",
  {"time_elapsed": 1234},
  {"response": 8, 8 //shows the response when target imaged appeared again//, null //shows that the target image did appear again but the participant didn't respond correctly, 8...}
//...
]

```

**Reply by vijaymarupudi** [reply]
You'll still need to use `clearInterval`, because after you finish the trial, it will continue to try to update an element that does not exist, creating errors in the process.

Now this will be the most complex part (well, a lot of the complexity has already been completed!). Let's aim for this format

```javascript
const trial_data = {
  responses: "", // ...
  time_elapsed: 23424,
  distractor_data: [
    { image: "target.png", response: 8, correct_response: true },
    //...
  ],
};
```

Unless it's absolutely necessary, the lack of a response should be handled in data processing in post, otherwise the code will become more complex. Let me know what you'd like.

The responses and time elapsed are handled by the other plugin that you've
combined. 

To collect distractor data, we'll need a state array for the data. Should initialize it before the `after_response` callback.

Note. You should pass the keyCode for the choices you'd like to the plugin instead of the string `'8'`, this will make the code easier.

Since the submit button listener uses the `distractor_data` array, I have moved it below the after response callback. I have also edited the submit function to include the distractor_data array.

I'm not fully sure if this will work, as your code isn't running directly from the git repository, I'm sure changes will be necessary.

Here is the entire diff

```diff
diff --git a/jspsych-survey-distractor.js b/jspsych-survey-distractor.js
index c0d5bb2..143233c 100644
--- a/jspsych-survey-distractor.js
+++ b/jspsych-survey-distractor.js
@@ -264,6 +264,23 @@ jsPsych.plugins["survey-distractor"] = (function () {
       key: null,
     };
 
+
+    // function to handle responses by the subject
+    
+    var distractor_data = []
+    var after_response = function (info) {
+      // after a valid response, the stimulus will have the CSS class 'responded'
+      // which can be used to provide visual feedback that a response was recorded
+      display_element_distractor.querySelector(
+        "#distractor-stimulus"
+      ).className += " responded";
+
+      // only record the choice response
+      if (trial.choices.includes(info.key)) {
+        distractor_data.push({ image: img.src, response: info.key, correct_response: img.src === 'target.png' })
+      }
+    };
+
     display_element_main_task
       .querySelector("#main_task-form")
       .addEventListener("submit", function (e) {
@@ -290,9 +307,8 @@ jsPsych.plugins["survey-distractor"] = (function () {
         // save data
         var trial_data = {
           responses: JSON.stringify(question_data),
-          rt: response.rt,
-          stimulus: trial.stimulus,
-          key_press: response.key,
+          time_elapsed: response.rt,
+          distractor_data: distractor_data
         };
 
         display_element.innerHTML = "";
@@ -301,20 +317,6 @@ jsPsych.plugins["survey-distractor"] = (function () {
         jsPsych.finishTrial(trial_data);
       });
 
-    // function to handle responses by the subject
-    var after_response = function (info) {
-      // after a valid response, the stimulus will have the CSS class 'responded'
-      // which can be used to provide visual feedback that a response was recorded
-      display_element_distractor.querySelector(
-        "#distractor-stimulus"
-      ).className += " responded";
-
-      // only record the choice response
-      if (response.key == trial.choices) {
-        response = info;
-      }
-    };
-
     // start the response listener
     if (trial.choices != jsPsych.NO_KEYS) {
       var keyboardListener = jsPsych.pluginAPI.getKeyboardResponse({
```

I've also included the changed file. I had to change the file extension because of Github rules, feel free to change it back!


[jspsych-survey-distractor.txt](https://github.com/jspsych/jsPsych/files/5016272/jspsych-survey-distractor.txt)

**Reply by avashipman** [reply]
Wow this is amazing! Thank you so so much!!! 

The distractor_data looks good as it is, so I will check the lack of response in data analysis. The distractor_data is working, but it only collects the data once. So when I press '8', it makes the whole 'distractor-stimulus' class 'responded' so every image that appears after I press '8' is considered responded already and the first image that got the response is the only one that is displayed in the distractor data at the end.

I have tried taking away the class 'responded' and making `info.key = trial.choices`, but the problem persists. I am thinking it may be because the id = 'distractor-stimulus' includes all the images so having the class 'responded' correspond with that id, making the response apply to all the images instead of the singular image. I tried making the [img](https://github.com/avashipman/Survey-Distractor/blob/bd3663ca241f26eb2bb119bbde66bb48c3952256/jspsych-survey-distractor.js#L251) = trail.stimulus and adjust the code accordingly,  but it just produces the array and not the individual images so nothing appears on the screen. I also tried taking out the id entirely to see what happens and the same thing as trial.stimulus occurs.

Same concept with the img tag at [L234-L247](https://github.com/avashipman/Survey-Distractor/blob/bd3663ca241f26eb2bb119bbde66bb48c3952256/jspsych-survey-distractor.js#L234-L247). It may be that since all the images are in one img tag, the response applies to all the images. I am not sure if this is the problem though.

**Reply by vijaymarupudi** [reply]
I think it's important to distinguish without the image source `img.src`, which are the different urls you want to show the participant, with the image tag `<img>`, which you have only one of cause you are showing one image at a time.

The problem here is that `jsPsych.pluginAPI.getKeyboardResponse` takes an argument called persist, which was previous false. We want the keyboard responses to persist, so I changed that to `true`. Since it'll persist, like `clearInterval`, I'll have to remove the listener before I finish the trial.

Additionally, you want the `responded` class to reset when the image changes. This can be easily done in the setInterval callback function.

Here's the diff of the changes.

```diff
diff --git a/jspsych-survey-distractor.js b/jspsych-survey-distractor.js
index fe054eb..2e22d3c 100644
--- a/jspsych-survey-distractor.js
+++ b/jspsych-survey-distractor.js
@@ -254,6 +254,7 @@ jsPsych.plugins["survey-distractor"] = (function () {
       console.log(img);
       index++;
       index = index % trial.stimulus.length;
+      img.classList.remove('responded')
       img.src = trial.stimulus[index];
     }
     const intervalId = setInterval(changeImage, trial.response_next_image);
@@ -315,6 +316,7 @@ jsPsych.plugins["survey-distractor"] = (function () {
 
         // next trial
         clearInterval(intervalId)
+        jsPsych.pluginAPI.cancelAllKeyboardResponses();
         jsPsych.finishTrial(trial_data);
       });
 
@@ -323,7 +325,7 @@ jsPsych.plugins["survey-distractor"] = (function () {
       var keyboardListener = jsPsych.pluginAPI.getKeyboardResponse({
         callback_function: after_response,
         valid_responses: trial.choices,
-        persist: false,
+        persist: true,
         allow_held_key: false,
       });
       keyboardListener;
```

[jspsych-survey-distractor.txt](https://github.com/jspsych/jsPsych/files/5019082/jspsych-survey-distractor.txt)

**Reply by avashipman** [reply]
This makes sense and worked perfectly! Your explanations were informative and extremely helpful. I can't say thank you enough for all of your help and time!! You're awesome!!!

**Reply by vijaymarupudi** [reply]
You're welcome! Glad to have helped :)
